twister: extend verification of test status by checking ztest suite name

gopiotr opened the pull request:
Test wirtten in ztest framework should print in output information about test suite name defined in test's source code (as a first argument of ztest_test_suite() function). Changes proposed in this PR make it possible to find such
test names in c files. Next they are used during test execution to verify if performed test was performed properly (verification can base not only on existance of "PROJECT EXECUTION SUCCESSFUL" info). 

So for example if test suites definition looks like this in file `tests/lib/gui/lvgl/src/main.c`:
```
ztest_test_suite(lvgl_screen, ztest_unit_test(test_get_default_screen),
	ztest_unit_test(test_add_delete_screen));
ztest_test_suite(lvgl_fs, ztest_user_unit_test_setup_teardown(
		test_add_img, setup_fs, teardown_fs));
```
Then Twister after parsing this file it will save defined ztest test suite names as `["lvgl_screen", "lvgl_fs"]`. Next during test execution harness will try to detect `"Running test suite (?P<suite_name>.*)"` regex in output, and save detected `suite_name` in `detected_suite_names` list. 

At the end if test has status pass and if expected test suite names list is not empty there is additionally verification if detected suite names correspond to expected ones.

I also tested this solution in situation when test suites are defined like this in file `tests/subsys/logging/log_core_additional/src/log_test.c` with `#ifdef` macro:
```
#ifdef CONFIG_USERSPACE
	ztest_test_suite(test_log_core_additional,
			 ztest_user_unit_test(test_log_from_user),
			 ztest_user_unit_test(test_log_hexdump_from_user),
			 ztest_user_unit_test(test_log_generic_user),
			 ztest_user_unit_test(test_log_filter_set),
			 ztest_user_unit_test(test_log_panic),
			 ztest_user_unit_test(test_log_msg2_create_user));
	ztest_run_test_suite(test_log_core_additional);
#else
	ztest_test_suite(test_log_core_additional,
			 ztest_unit_test(test_multiple_backends),
			 ztest_unit_test(test_log_generic),
			 ztest_unit_test(test_log_domain_id),
			 ztest_unit_test(test_log_severity),
			 ztest_unit_test(test_log_timestamping),
			 ztest_unit_test(test_log_early_logging),
			 ztest_unit_test(test_log_sync),
			 ztest_unit_test(test_log_thread),
			 ztest_unit_test(test_log_msg2_create)
			 );
	ztest_run_test_suite(test_log_core_additional);
#endif
```
In this situation c file parser will detect two test suites (they can have not the same name), but during execution test only one of them will be performed. Due to that fact I do not verify if all test suite names detected in c file will occur in output, but I rather verify if printed test suite name occur in expected suite names. I know that this is not ideal solution, but it will help to avoid situation when some test suite is intentionally skipped (by `#ifdef` macro) and even so Twister will assign to this test fail status. I described it also in added `_verify_ztest_suite_name` method's doc string.

As an effect of my changes situation described by @nvlsianpu in this issue: https://github.com/zephyrproject-rtos/zephyr/issues/42396 could be detected much more earlier.
With those proposed changes after run Twister by command:
```
scripts/twister --json-report -p native_posix -T tests/subsys/settings/functional/nvs/
```
The output from `native_posix` is like this:
```
2022-02-04 12:05:24,498 - twister - DEBUG - OUTPUT: *** Booting Zephyr OS build v3.0.0-rc1-234-g372a9f4918e5  ***
2022-02-04 12:05:24,499 - twister - DEBUG - OUTPUT: ===================================================================
2022-02-04 12:05:24,499 - twister - DEBUG - OUTPUT: PROJECT EXECUTION SUCCESSFUL
```
But the result in report is marked as fail with proper comment:
`twister.json` report:
```
{
    "testcase":"system.settings.functional.nvs.clear_settings",
    "arch":"posix",
    "platform":"native_posix",
    "status":"failed",
    "reason":"Test suite names were not printed or some of them in output do not correspond with expected: ['settings_test_suite']",
    "execution_time":2.007908821105957,
    "test_output":"*** Booting Zephyr OS build v3.0.0-rc1-234-g372a9f4918e5  ***\n===================================================================\nPROJECT EXECUTION SUCCESSFUL\n"
},
```
`twister_report.xml` report:
```
<testcase classname="system.settings" name="system.settings.functional.nvs.clear_settings" time="2.007909">
<failure type="failure" message="failed">
*** Booting Zephyr OS build v3.0.0-rc1-234-g372a9f4918e5 *** =================================================================== PROJECT EXECUTION SUCCESSFUL
</failure>
</testcase>
```

During adding this feature I also discover, that `scan_path` method in `TestCase` class in some cases wrongly choses `src` folder (with test source code). For example in those tests:

> tests/subsys/settings/fcb/raw
> tests/subsys/settings/littlefs/raw
> tests/subsys/settings/littlefs/base64
> tests/subsys/settings/nvs/raw
> tests/subsys/settings/functional/file
> tests/subsys/settings/functional/fcb
> tests/subsys/settings/functional/nvs

`src` folder is placed in "parent" directory (in context of `testcase.yaml` place). I add possibilities to detected such situation and correctly recognize `src` place. It should be discussed if in such situations occurrence of `src` directory should be unified, to be more predictable. Especially discussion about it should be conducted in case of this test (and its `src` directory placement): `tests/bluetooth/df` which also is problematic for Twister's c file parsing.

Moreover I also observed, that Twister is not prepared for parsing new ztest API available here:
`subsys/testsuite/ztest/src/ztest_new.c`
`subsys/testsuite/ztest/include/ztest_test_new.h`
I added special method for proper parsing of this API (`ZTEST_SUITE`, `ZTEST` and `ZTEST_F`)

gopiotr committed:
twister: verify ztest suite name in output

Test wirtten in ztest framework should print in output information about
test suite name defined in test's source code (as a first argument of
ztest_test_suite() function). This changes make it possible to find such
test names in c files. Next they are used during test execution to
verify if performed test was performed properly (verification can base
not only on existance of "PROJECT EXECUTION SUCCESSFUL" info).

Signed-off-by: Piotr Golyzniak <piotr.golyzniak@nordicsemi.no>

gopiotr committed:
twister: extend searching src directory

Extend scope of searching src directory with defined test source code.
In some cases for optimization reason src direcotry is placed in parent
directory (in relation to testcase.yaml file placement). This changes
allow to detect such situations.

Signed-off-by: Piotr Golyzniak <piotr.golyzniak@nordicsemi.no>

gopiotr commented:
Tests fail due to detected regression - some of test cases defined c files do not begin from `test_` characters. I'm going to fix those tests and next I rerun CI in this PR.

hakehuang commented on scripts/pylib/twister/twisterlib.py:
```diff
@@ -609,6 +642,8 @@ def handle(self):
             self.instance.reason = "Timeout"
             self.add_missing_testscases(harness)
 
+        self._verify_ztest_suite_name(harness.state, harness.detected_suite_names, handler_time)
```
I would suggest this is controlled by an option. and such check will make twister coupled tightly with ztest. and better to implement in a plug in way.

gopiotr replied:
I decided to introduce this checking as obligatory due to the fact that it do not impact on tests not written with ztest (for example samples) and it increase only the quality assurance of performed tests. For example this feature  could avoid situation described here:
https://github.com/zephyrproject-rtos/zephyr/issues/42396
Moreover it help to fix this problem by detect all affected tests:
https://github.com/zephyrproject-rtos/zephyr/pull/42439#pullrequestreview-877694014

hakehuang replied:
this has a requirement suite name must print in test log, which match its defines in the yaml file. if we decided to do so, we need update the ztest and twister document for this constrains.

gopiotr replied:
But, I do not perform this checking basing on any name from yaml file. For verification I use only suite names declared in C file in `ztest_test_suite(...)` or `ZTEST_SUITE(...)` functions.

So I think, that my propositions do not demand any update in ztest or twister documentation. This checking concerns only tests written in Ztest test framework, which requires to pass test suite name to every created test suites.

gopiotr committed:
twister: add search ZTEST_SUITE, ZTEST and ZTEST_F

New ztest API is available in those places:
subsys/testsuite/ztest/src/ztest_new.c
subsys/testsuite/ztest/include/ztest_test_new.h

This requires also changes in Twister code during parsing c files to
find properly test suites and testcases.

Signed-off-by: Piotr Golyzniak <piotr.golyzniak@nordicsemi.no>

gopiotr commented:
@nashif @hakehuang @PerMac @enjiamai @chen-png @yperess 
Could you take look at this PR and make review? ;)

PerMac commented on scripts/pylib/twister/harness.py:
```diff
@@ -214,8 +215,15 @@ def pytest_run(self, log_file):
+    test_suite_start_pattern = r"Running test suite (?P<suite_name>.*)"
 
     def handle(self, line):
+        test_suite_match = re.search(self.test_suite_start_pattern, line)
+        if test_suite_match:
+            suite_name = test_suite_match.group("suite_name")
+            self.detected_suite_names.append(suite_name)
+            self.ztest = True
```
Maybe you can add a minor optimization here? I believe there can only be a single test suite name printed in a single run. We don't need to check each and every line for test suite's name. Maybe if it is once found the check can be skipped?

gopiotr replied:
Documentation allows to create multiple test suites in one test file:
https://docs.zephyrproject.org/latest/guides/test/ztest.html#quick-start-unit-testing
And indeed some tests have several test suites defined in one test file, like those:
https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/drivers/i2s/i2s_api/src/main.c
https://github.com/zephyrproject-rtos/zephyr/blob/main/tests/bluetooth/controller/ctrl_data_length_update/src/main.c

So I cannot stop verify test suite name after find first one.

PerMac commented on scripts/pylib/twister/twisterlib.py:
```diff
@@ -468,6 +471,40 @@ def _set_skip_reason(self, harness_state):
+        (and not in reverse).
+        """
+        expected_suite_names = self.instance.testcase.ztest_suite_names
+        harness_state_pass = "passed"
+        if not expected_suite_names or \
+                not harness_state == harness_state_pass:
```
I don't see `harness_state_pass` being used anywhere else. Why not using:
`not harness_state == "passed:` directly? The best would be to have a global var for "passed" so we are sure it is everywhere the same, but I see this is no the case in current implementation and `=="passed"` is heavily used in harness.py, so I won't request for this here.

gopiotr replied:
Fine - I changed it ;)

PerMac commented on scripts/pylib/twister/twisterlib.py:
```diff
@@ -468,6 +471,40 @@ def _set_skip_reason(self, harness_state):
+            return
+        if not detected_suite_names:
+            self._missing_suite_name(expected_suite_names, handler_time)
+        for detected_suite_name in detected_suite_names:
```
As above. Can there be multiple detected suite names?

gopiotr replied:
Yes, there can be multiple detected suite names.

npal-cy commented:
Hi Guys,

I faced with similar problem. My device cy8cproto_062_4343w is flashing by pyOCD.  During the flash pyOCD executes reset of device, which cause the print out of terminal from previous flashed test application. Twister captures this output and complete the test based on wrong terminal data. (log in attachment)
[twister.log](https://github.com/zephyrproject-rtos/zephyr/files/8252553/twister.log)

I will check the fix from this PR, and provide the feedback. 

Also I think, will be good to have the option to force 'erase device' before testcase execution (all terminal data must be skipped during this phase..).

nashif reviewed (changes requested):
the concept is sound, few issues found and some improvements suggested.

gopiotr commented on scripts/pylib/twister/twisterlib.py:
```diff
@@ -468,6 +471,39 @@ def _set_skip_reason(self, harness_state):
+        self.set_state("failed", handler_time)
+        for k in self.instance.testcase.cases:
+            self.instance.results[k] = "FAIL"
+        self.instance.reason = f"Test suite names were not printed or some " \
```
I reported problem with those tests here: https://github.com/zephyrproject-rtos/zephyr/issues/43816 and I need to wait for opinion of those tests' author how to fix it.

gopiotr replied:
Suitable changes to fix this bug are proposed in this PR: https://github.com/zephyrproject-rtos/zephyr/pull/43822

nashif replied:
still there is a problem with the output, this is too verbose and badly formatted.

gopiotr commented on scripts/pylib/twister/twisterlib.py:
```diff
@@ -626,6 +662,9 @@ def handle(self):
 
         self._set_skip_reason(harness.state)
 
+        if harness_import.get_imported_harness_class_name() == "Test":
```
Done

chen-png reviewed (dismissed):
LGTM.
just curious one point, twister doc says it supports multiple test suites, but twister only could match the first pair of ztest_test_suite and ztest_run_test_suite, parse them and get all subcases between them, for later pairs of test suite, it won't parse them, so the final testcase.subcases only include the cases got form the first test suite pair.
however, this doesn't matter, because the real subcases, it's got from the harness.test when check and parse test output, it doesn't influence final cases counting.

chen-png commented:
This also will fix some errors in the case as below:
the first test is passed, when running the second test, the test binary update failed due to flash error or some other reasons, so it will still load the first test binary to exectue, and twister only judges test by "PROJECT EXECUTION SUCCESSFUL" , thus the second test is judged as passed two, but int the final xml results, it may c**ause a mixed cases, testcase.name(second test) + subcase.name(first test).**
if we can add a more judgement using test suite name, then this issue will be fixed.

gopiotr commented:
> twister only could match the first pair of ztest_test_suite and ztest_run_test_suite, parse them and get all subcases between them, for later pairs of test suite, it won't parse them, so the final testcase.subcases only include the cases got form the first test suite pair.

@chen-png I totally agree with you, that this is problematic case at this moment. To fix it this method would have to be refactored:
https://github.com/zephyrproject-rtos/zephyr/blob/399ff8b356919e904ddb4bda16187078284e5220/scripts/pylib/twister/twisterlib.py#L1904
But it seems that it could be challenging task, because it demands analyzing new test suite "boundary" and moreover even if detection of all test cases will be performed, then detected testcases would have to be assigned to suitable test suite. "Luckily" this problem concerns a few tests, and as you mentioned - it do not impact on the final result, because testcases are detected also later during output parsing. 

Anyway, I'm not going to introduce this "extended testcase detection" especially in this PR, because it is already quite big and I think that proposed "testsuite name verification" is enough at this moment.

But all in all thank you for your feedback :)

nashif commented:
needs a rebase

gopiotr commented:
> needs a rebase

Done
